src/usb_hs.rs

```diff
@@ -0,0 +1,93 @@
+//! USB OTG peripherals
+//!
+//! Requires the `usb_hs` feature.
+//!
+//! Note that only full-speed mode is supported,
+//! external high-speed PHY is not supported.
+
+use crate::rcc;
+use crate::stm32;
+
+use crate::gpio::{
+    gpioa::{PA11, PA12},
+    AF10,
+};
+use crate::gpio::{
+    gpiob::{PB14, PB15},
+    Alternate, AF12,
+};
+
+use crate::time::Hertz;
+
+pub use synopsys_usb_otg::UsbBus;
+use synopsys_usb_otg::UsbPeripheral;
+
+pub struct USB1 {
+    pub usb_global: stm32::OTG1_HS_GLOBAL,
+    pub usb_device: stm32::OTG1_HS_DEVICE,
+    pub usb_pwrclk: stm32::OTG1_HS_PWRCLK,
+    pub pin_dm: PB14<Alternate<AF12>>,
+    pub pin_dp: PB15<Alternate<AF12>>,
+    pub prec: rcc::rec::Usb1Otg,
+    pub hclk: Hertz,
+}
+
+pub struct USB2 {
+    pub usb_global: stm32::OTG2_HS_GLOBAL,
+    pub usb_device: stm32::OTG2_HS_DEVICE,
+    pub usb_pwrclk: stm32::OTG2_HS_PWRCLK,
+    pub pin_dm: PA11<Alternate<AF10>>,
+    pub pin_dp: PA12<Alternate<AF10>>,
+    pub prec: rcc::rec::Usb2Otg,
+    pub hclk: Hertz,
+}
+
+macro_rules! usb_peripheral {
+    ($USB:ident, $GLOBAL:ident, $en:ident, $rst:ident) => {
+        unsafe impl Sync for $USB {}
+
+        unsafe impl UsbPeripheral for $USB {
+            const REGISTERS: *const () = stm32::$GLOBAL::ptr() as *const ();
+
+            const HIGH_SPEED: bool = true;
+            const FIFO_DEPTH_WORDS: usize = 1024;
+
+            const ENDPOINT_COUNT: usize = 9;
+
+            fn enable() {
+                let pwr = unsafe { &*stm32::PWR::ptr() };
+                let rcc = unsafe { &*stm32::RCC::ptr() };
+
+                cortex_m::interrupt::free(|_| {
+                    // USB Regulator in BYPASS mode
+                    pwr.cr3.modify(|_, w| w.usb33den().set_bit());
+
+                    // Enable USB peripheral
+                    rcc.ahb1enr.modify(|_, w| w.$en().set_bit());
+
+                    // Reset USB peripheral
+                    rcc.ahb1rstr.modify(|_, w| w.$rst().set_bit());
+                    rcc.ahb1rstr.modify(|_, w| w.$rst().clear_bit());
+                });
+            }
+
+            fn ahb_frequency_hz(&self) -> u32 {
+                // For correct operation, the AHB frequency should be higher
+                // than 30MHz. See RM0433 Rev 7. Section 57.4.4
+                assert!(self.hclk.0 > 30_000_000);
```

Comment:
BTW, this check is already present in https://github.com/stm32-rs/synopsys-usb-otg/blob/33dc554a8c2ce15e15f577b36d04de7ed3c85034/src/bus.rs#L282-L283

Reply:
Ok, thanks for finding that. For `HIGH_SPEED` = true, it actually uses the match statement above. Perhaps this is fine because we're only running the bus in Full-Speed mode anyhow?

Reply:
Oh, it's a mistake. The match statement should be executed for `HIGH_SPEED` = false. Actually, inside the driver from CubeIDE this TRDT value is calculated for enumeration speed, not for the maximum speed. As far as I understand, this >30MHz AHB frequency bound is suggested (in 57.4.4) to *guarantee* a correct operation, because you can't tell in advance what speed your device will enumerate into.